### PR TITLE
Feature/nrlf 563 sort immutable fields

### PR DESCRIPTION
**Pull request by andrewrae5-nhs**
added sorting to immutable fields, as per [NRLF-563](https://nhsd-jira.digital.nhs.uk/browse/NRLF-563)

### CODE REVIEW

File: api/producer/updateDocumentReference/src/v1/handler.py [hunk 1]
```diff
@@ -77,13 +77,36 @@ def document_pointer_exists(
...
+        key_values = list(key.items())[0][1]
+
+        if type(key_values) == dict:
+            return dict(sorted(key_values.items()))
+        elif type(key_values) == list:
+            keys_list = []
+            for sub_keys in key_values:
+                keys_list.extend(list(sub_keys.keys()))
+
+            return sorted(key_values, key=lambda item: sorted(keys_list)[0])
```
Comment (jaklinger) [inline api/producer/updateDocumentReference/src/v1/handler.py:94]: would the following not be sufficient (would sort any dictionary by keys then sort all values):

```suggestion
        return {k: _sort_key(key[k]) for k in sorted(key.keys())}
```

File: api/producer/updateDocumentReference/src/v1/handler.py [hunk 2]
```diff
@@ -77,13 +77,36 @@ def document_pointer_exists(
     )
 
 
+def _sort_key(key):
+
+    if type(key) == list:
+        return sorted(key)
```
Comment (jaklinger) [inline api/producer/updateDocumentReference/src/v1/handler.py:83]: does not work for list of dict, or a list of list, etc, i'd suggest:

```suggestion
        return sorted(key, key=_sort_key)
```

File: api/producer/updateDocumentReference/src/v1/handler.py
Comment (jaklinger) [inline api/producer/updateDocumentReference/src/v1/handler.py:1]: in other words, suggest to flatten function to:

```python
def _sort_key(key):
    if type(key) is list:
        return sorted(key, key=_sort_key)
    elif type(key) is dict:
        return [(k, _sort_key(key[k])) for k in sorted(key.keys())]
    return key
```

File: api/producer/updateDocumentReference/tests/unit/test_update_document_reference_v1_steps.py [hunk 3]
```diff
@@ -81,6 +82,134 @@ def test_compare_immutable_fields_success(mock__make_timestamp):
...
+            "custodian": {"S": "Y05868"},
+            "source": {"S": "NRLF"},
+            "type": {"S": "http://snomed.info/sct|736253002"},
+            "updated_on": {"S": "2022-10-25T15:47:49.732Z"},
+            "version": {"N": "1"},
+            "author": [
+                {"id": "a", "reference": "b"},
+                {"id": "b", "reference": "a"},
+                {"reference": "c", "display": "d"},
+            ],
```
Comment (jaklinger) [inline api/producer/updateDocumentReference/tests/unit/test_update_document_reference_v1_steps.py:167]: to fix the broken test:

```suggestion
```

Comment (andrewrae5-nhs) [reply api/producer/updateDocumentReference/tests/unit/test_update_document_reference_v1_steps.py]: whooops! strange, worked before :/

**Review by jaklinger — APPROVED** [review]

**Pull request merged**